core/imds/src/main/java/software/amazon/awssdk/imds/internal/DefaultEc2MetadataAsyncClient.java

```diff
@@ -0,0 +1,287 @@
+            return scheduledRetryAttempt(() -> get(path, newContext), newContext);
+        }).thenCompose(Function.identity()); // only java 12 has .exceptionallyCompose()
+    }
+
+    private CompletableFuture<String> sendAsyncRequest(SdkHttpFullRequest baseRequest) {
+        StringAsyncResponseHandler responseHandler = new StringAsyncResponseHandler();
+        SdkHttpContentPublisher requestContentPublisher = new SimpleHttpContentPublisher(baseRequest);
+        AsyncExecuteRequest metadataRequest = AsyncExecuteRequest.builder()
+                                                                 .request(baseRequest)
+                                                                 .requestContentPublisher(requestContentPublisher)
+                                                                 .responseHandler(responseHandler)
+                                                                 .build();
+        httpClient.execute(metadataRequest);
+        return responseHandler.getResult();
+    }
+
+    private CompletableFuture<MetadataResponse> scheduledRetryAttempt(Supplier<CompletableFuture<MetadataResponse>> supplier,
+                                                                      RetryPolicyContext retryPolicyContext) {
+        Duration retryDelay = retryPolicy.backoffStrategy().computeDelayBeforeNextRetry(retryPolicyContext);
+        Executor retryExecutor = retryAttempt -> asyncRetryScheduler.schedule(retryAttempt, retryDelay.toMillis(),
+                                                                              TimeUnit.MILLISECONDS);
+        return CompletableFuture.supplyAsync(supplier, retryExecutor).thenCompose(Function.identity());
+    }
+
+    private boolean shouldRetry(RetryPolicyContext retryPolicyContext, Throwable error) {
+        boolean maxAttemptReached = retryPolicyContext.retriesAttempted() >= retryPolicy.numRetries();
+        if (maxAttemptReached) {
+            return false;
+        }
+        return error instanceof RetryableException || error.getCause() instanceof RetryableException;
+    }
+
+    @Override
+    public Builder toBuilder() {
+        return builder()
+            .endpoint(this.endpoint)
+            .endpointMode(this.endpointMode)
+            .httpClient(this.httpClient)
+            .retryPolicy(this.retryPolicy)
+            .tokenTtl(this.tokenTtl);
+    }
+
+    @Override
+    public void close() {
+        httpClient.close();
+    }
+
+    private static final class StringResponseHandler implements HttpResponseHandler<String> {
+        @Override
+        public String handle(SdkHttpFullResponse response, ExecutionAttributes executionAttributes) throws Exception {
+            return IoUtils.toUtf8String(response.content().orElseThrow(() ->
+                SdkClientException.create("Unexpected error: empty response content")));
+        }
+    }
+
+    private static final class StringAsyncResponseHandler implements TransformingAsyncResponseHandler<String> {
```

Comment:
Do we actually need this class? Can't we check the status code in `StringResponseHandler`?

Reply:
We need it to build an `AsyncExecuteRequest` instance

Reply:
Ok, I got it, removing AsyncHandler class